### ISSUE TITLE
Non-deterministic errors running 'db.lua'

### ISSUE THREAD

**Issue by ptersilie** [ykjit/yk#761]
When running `tests/db.lua` using yk, I'm getting a few different errors:

- `malloc(): mismatching next->prev_size (unsorted)`
- `Segfault`
- `realloc(): invalid old size`

The errors always happen at the end of the program, just before exiting, which seems to suggest it's a problem when dropping/tearing down stuff. I've reverted all the way back to 5 June thinking some of our recent major changes where responsible. But these errors appear to go even further.

**Comment by ltratt** [comment]
Here's a possible cause: when we deopt, do we properly unwind the stack? For example, if we replace Rust frames on the stack, we should run the relevant `drop` methods. If there's C++ code, it may require us running some sort of exception code?

**Comment by vext01** [comment]
> For example, if we replace Rust frames on the stack

I've wondered about this, but Lukas' PR that replaces the Rust stack was only merged this week and he's saying this is a very old problem.

**Comment by ltratt** [comment]
I wonder if we can bisect this? There again, we can't go back very far with ease: as soon as we go back before the "yk build ykllvm" stuff was merged and fully working (roughly mid-May) building things becomes torturous. So this is quite probably a dead-end route.

**Comment by vext01** [comment]
@ptersilie did some bisecting yesterday, but I don't know the outcome.

**Comment by ltratt** [comment]
@vext01 He said earlier:

> I've reverted all the way back to 5 June thinking some of our recent major changes where responsible. But these errors appear to go even further.

**Comment by vext01** [comment]
yeah, but did he go further back than that even? I guess you can't go beyond the build system revamp...

**Comment by ltratt** [comment]
If he said June 5th, I'd be inclined to believe him!
